LLama/Native/Load/NativeLibraryConfig.cs

```diff
@@ -0,0 +1,605 @@
+
+        internal NativeLibraryName NativeLibraryName { get; }
+
+        internal NativeLogConfig.LLamaLogCallback? LogCallback { get; private set; } = null;
+
+        private void ThrowIfLoaded()
+        {
+            if (LibraryHasLoaded)
+                throw new InvalidOperationException("The library has already loaded, you can't change the configurations. " +
+                    "Please finish the configuration setting before any call to LLamaSharp native APIs." +
+                    "Please use NativeLibraryConfig.DryRun if you want to see whether it's loaded " +
+                    "successfully but still have chance to modify the configurations.");
+        }
+
+        /// <summary>
+        /// Set the log callback that will be used for all llama.cpp log messages
+        /// </summary>
+        /// <param name="callback"></param>
+        /// <exception cref="NotImplementedException"></exception>
+        public NativeLibraryConfig WithLogCallback(NativeLogConfig.LLamaLogCallback? callback)
+        {
+            ThrowIfLoaded();
+
+            LogCallback = callback;
+            return this;
+        }
+
+        /// <summary>
+        /// Set the log callback that will be used for all llama.cpp log messages
+        /// </summary>
+        /// <param name="logger"></param>
+        /// <exception cref="NotImplementedException"></exception>
+        public NativeLibraryConfig WithLogCallback(ILogger? logger)
+        {
+            ThrowIfLoaded();
+
+            // Redirect to llama_log_set. This will wrap the logger in a delegate and bind that as the log callback instead.
+            NativeLogConfig.llama_log_set(logger);
+
+            return this;
+        }
+
+        /// <summary>
+        /// Try to load the native library with the current configurations, 
+        /// but do not actually set it to <see cref="NativeApi"/>.
+        /// 
+        /// You can still modify the configuration after this calling but only before any call from <see cref="NativeApi"/>.
+        /// </summary>
+        /// <returns>Whether the running is successful.</returns>
+        public bool DryRun()
+        {
+            LogCallback?.Invoke(LLamaLogLevel.Debug, $"Beginning dry run for {this.NativeLibraryName.GetLibraryName()}...");
+            return NativeLibraryUtils.TryLoadLibrary(this) != IntPtr.Zero;
+        }
+    }
+
+    /// <summary>
+    /// A class to set same configurations to multiple libraries at the same time.
+    /// </summary>
+    public sealed class NativeLibraryConfigContainer
+    {
+        private NativeLibraryConfig[] _configs;
+
+        /// <summary>
+        /// All the configurations in this container.
+        /// Please avoid calling this property explicitly, use <see cref="NativeLibraryConfig.LLama"/>
+        /// and <see cref="NativeLibraryConfig.LLavaShared"/> instead.
+        /// </summary>
+        public NativeLibraryConfig[] Configs => _configs;
```

Comment:
Can this be made `internal` if people aren't meant to access it directly?

Reply:
I exposed it because it's needed to keep auto-download in another package. I could remove it and hardcode it in #692